File: common/src/main/java/com/wolfyscript/utilities/gui/WindowImpl.java
package com.wolfyscript.utilities.gui;

import com.google.common.base.Preconditions;
import com.wolfyscript.utilities.KeyedStaticId;
import com.wolfyscript.utilities.WolfyUtils;
import com.wolfyscript.utilities.gui.callback.InteractionCallback;
import com.wolfyscript.utilities.gui.functions.SerializableSupplier;
import com.wolfyscript.utilities.platform.scheduler.Task;
import com.wolfyscript.utilities.tuple.Pair;

import java.util.*;

@KeyedStaticId(key = "window")
public final class WindowImpl implements Window {

    private final String id;
    private final Router router;
    private final WolfyUtils wolfyUtils;
    private final Integer size;
    private final WindowType type;
    private String staticTitle = null;
    private final SerializableSupplier<net.kyori.adventure.text.Component> dynamicTitle;
    private final InteractionCallback interactionCallback;
    final Collection<Component> componentsToRender;

    // Intervalls
    final List<Pair<Runnable, Long>> intervalRunnables = new ArrayList<>();
    final List<Task> intervalTasks = new ArrayList<>();

    WindowImpl(String id,
               Router router,
               Integer size,
               WindowType type,
               String staticTitle,
               SerializableSupplier<net.kyori.adventure.text.Component> dynamicTitle,
               InteractionCallback interactionCallback,
               Collection<Component> components) {
        Preconditions.checkNotNull(id);
        Preconditions.checkNotNull(interactionCallback);
        Preconditions.checkArgument(size != null || type != null, "Either type or size must be specified!");
        this.id = id;
        this.router = router;
        this.wolfyUtils = router.getWolfyUtils();
        this.size = size;
        this.type = type;
        this.staticTitle = staticTitle;
        this.interactionCallback = interactionCallback;
        this.componentsToRender = components;
        this.dynamicTitle = dynamicTitle;
    }

    @Override
    public Window construct(GuiHolder holder, ViewRuntime viewManager) {
        return this;
    }

    @Override
    public void open(ViewRuntime viewRuntime) {
        for (Task intervalTask : intervalTasks) {
            intervalTask.cancel();
        }
        intervalTasks.clear();
        for (Pair<Runnable, Long> intervalRunnable : intervalRunnables) {
            Task task = wolfyUtils.getCore().platform().scheduler().task(wolfyUtils)
                    .interval(intervalRunnable.getValue())
                    .delay(1).execute(intervalRunnable.getKey()).build();
            intervalTasks.add(task);
        }
    }

    @Override
    public void close(ViewRuntime viewRuntime) {
        for (Task intervalTask : intervalTasks) {
            intervalTask.cancel();
        }
        intervalTasks.clear();
    }

    @Override
    public void render(GuiHolder guiHolder, ViewRuntime viewManager, RenderContext context) {
        if (dynamicTitle != null) {
            context.updateTitle(guiHolder, dynamicTitle.get());
        }

        for (Component component : componentsToRender) {
            var position = component.position();
            if (position == null) continue;
            ((ViewRuntimeImpl) guiHolder.getViewManager()).updateLeaveNodes(component, position.slot());
            context.enterNode(component);
            component.render(viewManager, guiHolder, context);
            context.exitNode();
        }
    }

    @Override
    public WolfyUtils getWolfyUtils() {
        return wolfyUtils;
    }

    @Override
    public String getID() {
        return id;
    }

    @Override
    public Router router() {
        return router;
    }

    @Override
    public InteractionResult interact(GuiHolder holder, InteractionDetails interactionDetails) {
        return null;
    }

    @Override
    public InteractionCallback interactCallback() {
        return interactionCallback;
    }

    @Override
    public Set<? extends Component> childComponents() {
        return Set.of();
    }

    @Override
    public Optional<Component> getChild(String id) {
        return Optional.empty();
    }

    @Override
    public Optional<Integer> getSize() {
        return Optional.ofNullable(size);
    }

    @Override
    public Optional<WindowType> getType() {
        return Optional.ofNullable(type);
    }

    @Override
    public net.kyori.adventure.text.Component createTitle(GuiHolder holder) {
        return wolfyUtils.getChat().getMiniMessage().deserialize(staticTitle);
    }

    public String getStaticTitle() {
        return staticTitle;
    }

    @Override
    public int width() {
        return size / height();
    }

    @Override
    public int height() {
        return size / 9;
    }

}

File: common/src/main/java/com/wolfyscript/utilities/gui/WindowImpl.kt
package com.wolfyscript.utilities.gui

import com.fasterxml.jackson.annotation.JacksonInject
import com.fasterxml.jackson.annotation.JsonProperty
import com.google.common.base.Preconditions
import com.wolfyscript.utilities.KeyedStaticId
import com.wolfyscript.utilities.WolfyUtils
import com.wolfyscript.utilities.functions.ReceiverConsumer
import com.wolfyscript.utilities.functions.ReceiverFunction
import com.wolfyscript.utilities.gui.callback.TextInputCallback
import com.wolfyscript.utilities.gui.callback.TextInputTabCompleteCallback
import com.wolfyscript.utilities.gui.reactivity.ReactiveSource
import com.wolfyscript.utilities.gui.router.Router
import com.wolfyscript.utilities.gui.router.RouterImpl
import com.wolfyscript.utilities.platform.scheduler.Task
import com.wolfyscript.utilities.tuple.Pair
import net.kyori.adventure.text.Component

@KeyedStaticId(key = "window")
class WindowImpl internal constructor(
    @JsonProperty("id") override val id: String,
    @JsonProperty("size") override var size: Int?,
    @JsonProperty("type") override val type: WindowType? = null,
    @JacksonInject("wolfyutils") override val wolfyUtils: WolfyUtils,
    @JacksonInject("context") private val context: BuildContext,
) :
    Window,
    ReactiveSource by context.reactiveSource {
    override var title: Component? = null
    override val router: Router = RouterImpl(wolfyUtils, context, this)
    override var resourcePath: String? = null

    var textInputCallback: TextInputCallback? = null
    var textInputTabCompleteCallback: TextInputTabCompleteCallback? = null

    // Intervalls
    val intervalRunnables: List<Pair<Runnable, Long>> = ArrayList()
    val intervalTasks: MutableList<Task> = ArrayList()

    init {
        Preconditions.checkArgument(size != null || type != null, "Either type or size must be specified!")
    }

    override fun title(titleUpdate: ReceiverFunction<Component?, Component?>) {
        context.reactiveSource.createEffect<Unit> {
            title = with(titleUpdate) { title.apply() }

            (context.runtime as ViewRuntimeImpl).renderer.updateTitle(title)
        }
    }

    override fun open() {
        for (intervalTask in intervalTasks) {
            intervalTask.cancel()
        }
        intervalTasks.clear()
        for (intervalRunnable in intervalRunnables) {
            val task = wolfyUtils.core.platform.scheduler.task(wolfyUtils)
                .interval(intervalRunnable.value)
                .delay(1).execute(intervalRunnable.key).build()
            intervalTasks.add(task)
        }
        router.open()
    }

    override fun close() {
        for (intervalTask in intervalTasks) {
            intervalTask.cancel()
        }
        intervalTasks.clear()

//        (context.runtime as ViewRuntimeImpl).renderingGraph.removeNode(0)
    }

    override fun routes(routerConfiguration: ReceiverConsumer<Router>) {
        with(routerConfiguration) { router.consume() }
    }

    override fun width(): Int {
        return size?.div(height()) ?: 9
    }

    override fun height(): Int {
        return size?.div(9) ?: 1
    }

    override fun onTextInput(inputCallback: TextInputCallback?) {
        this.textInputCallback = inputCallback
    }

    override fun onTextInputTabComplete(textInputTabCompleteCallback: TextInputTabCompleteCallback?) {
        this.textInputTabCompleteCallback = textInputTabCompleteCallback
    }

}
